File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/CertificationAuthorityAttributeInfo.kt
package com.jetbrains.signatureverifier.serialization

import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.cms.Attribute
import org.bouncycastle.asn1.x509.AlgorithmIdentifier

@Serializable
data class CertificationAuthorityAttributeInfo(
  val identifier: TextualInfo,
  val content: List<AlgorithmInfo>
) : AttributeInfo {

  constructor(attribute: Attribute) : this(
    TextualInfo.getInstance(attribute.attrType),
    attribute.attributeValues.map { AlgorithmInfo(AlgorithmIdentifier.getInstance(it)) }
  )

  override fun toAttributeDLSequence(): DLSequence = listOf(
    identifier.toPrimitive(),
    content.toPrimitiveList().toDLSet()
  ).toDLSequence()
}

File: net/JetBrains.Serialization/src/Dataholders/AttributeInfos/CertificationAuthorityAttributeInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;
using JetBrains.Serialization;
using Org.BouncyCastle.Asn1.X509;

[JsonObject(MemberSerialization.OptIn)]
public class CertificationAuthorityAttributeInfo : AttributeInfo
{
  [JsonProperty("Identifier")] protected override TextualInfo Identifier { get; }
  [JsonProperty("Content")] private List<AlgorithmInfo> _content;

  [JsonConstructor]
  public CertificationAuthorityAttributeInfo(TextualInfo identifier, List<AlgorithmInfo> content)
  {
    Identifier = identifier;
    _content = content;
  }

  public CertificationAuthorityAttributeInfo(Attribute attribute)
  {
    Identifier = TextualInfo.GetInstance(attribute.AttrType);
    _content = attribute.AttrValues.ToArray().Select(item => new AlgorithmInfo(AlgorithmIdentifier.GetInstance(item)))
      .ToList();
  }

  public override Asn1Encodable GetPrimitiveContent() => _content.ToPrimitiveDerSet();
}